src/RobotsTxt.php

```diff
@@ -32,34 +32,68 @@ public static function create(string $source): self
         return new self($source);
     }
 
+    // Google says (https://developers.google.com/search/reference/robots_txt) :
+    // Only one group of group-member records is valid for a particular crawler.
+    // The crawler must determine the correct group of records by finding the group
+    // with the most specific user-agent that still matches.
     public function allows(string $url, ?string $userAgent = '*'): bool
     {
+        $userAgent = strtolower($userAgent);
+
+        if ($userAgent === null) {
+            $userAgent = '*';
+        }
+
         $path = parse_url($url, PHP_URL_PATH) ?? '';
 
-        $disallows = $this->disallowsPerUserAgent[$userAgent] ?? $this->disallowsPerUserAgent['*'] ?? [];
+        if ($userAgent != '*' && isset($this->disallowsPerUserAgent[$userAgent])) {
+            return ! $this->pathIsDenied($path, $this->disallowsPerUserAgent[$userAgent]) ?? true;
+        } elseif ($userAgent != '*' && $wildCardUserAgent = $this->getWildCardUserAgent($userAgent)) {
+            return ! $this->pathIsDenied($path, $this->disallowsPerUserAgent[$wildCardUserAgent]) ?? true;
+        } elseif (isset($this->disallowsPerUserAgent['*'])) {
+            return ! $this->pathIsDenied($path, $this->disallowsPerUserAgent['*']) ?? true;
+        }
 
-        return ! $this->pathIsDenied($path, $disallows);
+        return true;
     }
 
-    protected function pathIsDenied(string $path, array $disallows): bool
+    protected function getWildCardUserAgent(string $userAgent): ?string
     {
-        foreach ($disallows as $disallow) {
-            $trimmedDisallow = rtrim($disallow, '/');
-
-            if (in_array($path, [$disallow, $trimmedDisallow])) {
-                return true;
+        if ($userAgent !== '*') {
+            for ($i = 1; $i <= strlen($userAgent); $i++) {
+                $wildCardUserAgent = substr($userAgent, 0, $i).'*';
+                if (isset($this->disallowsPerUserAgent[$wildCardUserAgent])) {
+                    return $wildCardUserAgent;
+                }
             }
+        }
 
-            if (! $this->concernsDirectory($disallow)) {
-                continue;
-            }
+        return null;
+    }
 
-            if ($this->isUrlInDirectory($path, $disallow)) {
-                return true;
+    protected function pathIsDenied(string $path, array $rules)
+    {
+        foreach ($rules as $pattern => $rule) {
+            if ($this->match($pattern, $path)) {
+                return $rule;
             }
         }
+    }
+
+    protected function complexRule($path): boolean
+    {
+        return strpos($path, ['$', '*']);
+    }
+
+    protected function match($pattern, $string)
+    {
+        $pattern = preg_quote($pattern, '/');
```

Comment:
I'd like to avoid regular expressions here. Could you replace this with regular string functions?

Reply:
How ?